ibis/common/patterns.py

```diff
@@ -585,15 +370,16 @@ class Replace(Slotted, Pattern):
     builder: Builder
 
     def __init__(self, matcher, replacer):
-        super().__init__(pattern=pattern(matcher), builder=builder(replacer))
+        builder = deferred(replacer).__builder__
+        super().__init__(pattern=pattern(matcher), builder=builder)
 
     def match(self, value, context):
         value = self.pattern.match(value, context)
         if value is NoMatch:
             return NoMatch
         # use the `_` reserved variable to record the value being replaced
         # in the context, so that it can be used in the replacer pattern
-        context[_] = value
+        context["_"] = value
```

Comment:
Chose to reserve `"_"` string key since `var("_")` now produces a `Deferred` object wrapping a `Variable("_")` where the wrapped variable is hashable but the outer deferred object is not, so it is not possible to pass the context as `{_: <value>}` anymore.